Integrate forked Capstan

miha-plesko opened the pull request:
Added support for forked Capstan tool. User can now create OSv
unikernels using precompiled packages.

There are three configuration files available:
* $PROJECT_ROOT/meta/run.yaml (application structure description)
* $PROJECT_ROOT/meta/package.yaml (list of required packages)
* $PROJECT_ROOT/manifest.yaml (additional parameters for Capstan)
Only the first one is required, other two are optional.

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

miha-plesko commented:
Dear @ilackarms, dear @gberginc, I've managed to prepare the forked Capstan integration at last! :champagne: 

I realize that the PR is quite large, but I think it's easiest to show you the whole story at once. I've followed the steps that we've specified on the videocall so you should probably be able to understand the code.

I'm looking forward to your comments.

PS: Below please find examples for each of the languages that the forked Capstan provides:
[forked-capstan-examples.zip](https://github.com/emc-advanced-dev/unik/files/670968/forked-capstan-examples.zip)

miha-plesko commented:
He-yo, 2017 up-and-running, hope you're all doing well back there at EMC! 😄
Does this look somewhat what we wanted to have?

ilevine commented:
We have been busy with the holiday and the new year. We will get to reviewing this PR soon. Sorry for the delay

ilevine commented:
* Please make sure to add appropriate targets to the Makefile for the new container/compiler you've added to unik
* Having a large set of changes all contained in a single commit makes it difficult for us to reason through the code. We strongly suggest splitting future changes into smaller commits (such that each commit allows unik to compile), where the new feature is integrated into the cli/daemon only in the last commit.

Commit:
Update documentation with forked Capstan description

Added description about how to use the three supported
languages for osv base (java, nodejs and native). Documentation
for Java language had to be rewriten since we no longer take
part in user's application source compilation (with mvn). We assume
that user has compiled her application PRIOR deploying it into
unikernel.

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

Commit:
Add compilers-osv-dynamic container

This container contains forked Capstan and is able
to compose arbitrary OSv image. It replaces the
compilers-osv-java that used the old Capstan.

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

Commit:
Additional section in compiler's usage string

Turns out that compiler description should start with
short general description. This way the usage string
reads better.

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

Commit:
Add OSv dynamic compiler and QEMU finisher

OSv dynamic compiler is not exposed to the CLI directly, but
other compilers will interact with forked Capstan thru
it. Also added support for QEMU here since images composed
with Capstan are actually QEMU images.

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

Commit:
Expose the three runtimes to the CLI

At the moment, forked Capstan supports three runtimes:
Java, NodeJS and Native. They correspond to UniK's term
'language'. So this commit adds support for:
```
unik build --base osv --language java
unik build --base osv --language nodejs
unik build --base osv --language native
```
With 'native' we usually refer to compiled C/C++, but it
also allows user to run arbitrary application package from
the public repository.

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

Commit:
OpenStack provider now considers logical size to pick flavor

When building unikernel with forked Capstan, user can specify
its filesystem size (10 GB by default). Actual size of the
unikernel is usually small, but can grow when the unikernel is
running (up to the filesystem size). On OpenStack we must run
pick flavor large enough to fit the filesystem size of the unikernel.

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

miha-plesko commented:
Thanks for the response, @ilevine. I've split the PR into 6 smaller commits and added Makefile target for the new container.

As you can see, I've taken liberty to fully replace existing OSv integration with the new one (so that only one implementation of Capstan will be used once we merge this). I'm looking forward to be hearing your opinion on this. Should it bother you too much, I'll try to find a way to keep both integrations.

gberginc commented on pkg/providers/openstack/stage.go:
```diff
@@ -54,20 +54,22 @@ func (p *OpenstackProvider) Stage(params types.StageImageParams) (_ *types.Image
-	imageSizeB := rawImageFile.Size()
-	imageSizeMB := int(unikos.Bytes(imageSizeB).ToMegaBytes())
+	// Image size - use logical size unless physical size is bigger
+	imagePhysicalSizeMB := int(unikos.Bytes(rawImageFile.Size()).ToMegaBytes())
```
Why did you have to rename the variable? This line does the same as the two above, doesn't it? Also, I don't think it's safe to remove the TODO as you haven't actually read the logical image size.

miha-plesko replied:
I've dropped this change and has scheduled it for another PR.

gberginc commented on pkg/providers/openstack/stage.go:
```diff
@@ -54,20 +54,22 @@ func (p *OpenstackProvider) Stage(params types.StageImageParams) (_ *types.Image
 
 	// Pick flavor.
-	flavor, err := pickFlavor(clientNova, imageSizeMB, 0)
+	flavor, err := pickFlavor(clientNova, params.RawImage.RunSpec.MinInstanceDiskMB, 0)
```
Do you think it would make sense to store the actual volume size defined in the flavor? It can actually be larger than the one we requested.

miha-plesko replied:
Yes, it would make sense. But it cannot be done easily. UniK separates compilers and providers very strictly so at the time we are building unikernel (with Capstan), UniK is not aware of available flavors yet.

I'd leave this for now. With manifest.yaml we allow user to set arbitrary logical size and at this point I'd leave it to user to pick the size that matches flavor.

gberginc commented on pkg/providers/openstack/stage.go:
```diff
@@ -81,7 +83,7 @@ func (p *OpenstackProvider) Stage(params types.StageImageParams) (_ *types.Image
 		RunSpec:        params.RawImage.RunSpec,
 		StageSpec:      params.RawImage.StageSpec,
-		SizeMb:         int64(imageSizeMB),
+		SizeMb:         int64(imagePhysicalSizeMB),
```
Agani, should we store the flavor disk size?

miha-plesko replied:
Please see comment above.

gberginc commented on pkg/types/types.go:
```diff
@@ -134,11 +134,13 @@ const (
-	StorageDriver         StorageDriver      `json:"StorageDriver,omitempty"`
-	VsphereNetworkType    VsphereNetworkType `json:"VsphereNetworkType"`
-	Compiler              string             `json:"Compiler,omitempty"`
+	DefaultInstanceMemory int `json:"DefaultInstanceMemory"` //required for all compilers
```
Can you shorten the comment here? I really don't like the fact that fields are misaligned just because of the comment.

miha-plesko replied:
This change was dropped from the PR and is scheduled for the next one.

gberginc reviewed (changes requested):
I would move this into a separate PR as it does not necessarily relate to the other changes.

The size is currently only required by the OpenStack provider - I am assuming that other providers are still able to work without the use of this additional option?

ilevine commented on pkg/compilers/osv/osv_dynamic.go:
```diff
@@ -0,0 +1,76 @@
+}
+
+// readImageSizeFromManifestMB parses manifest.yaml and returns image size in MegaBytes
+func readImageSizeFromManifestMB(projectDir string) unikos.MegaBytes {
```
the error should be checked here

miha-plesko replied:
Merged this function with the upper one so the problem no longer exist.

ilevine commented on containers/compilers/osv/java-compiler/java-main-caller/src/main/java/com/emc/wrapper/UDPBootstrap.java:
```diff
@@ -1,160 +0,0 @@
-package com.emc.wrapper;
```
can you explain the purpose of removing the OSv Wrapper?

gberginc replied:
Since the modified capstan now integrates java a bit differently, the initial idea was to remove the current version completely. We discussed yesterday with @miha-plesko about this PR and decided to abandon it an split into several smaller PRs, incrementally introducing the new capabilities for OSv.

One of the things we decided was to stick with current support for Java and add just new support for Node and "native". This will facilitate discussion about the proposed benefits, as well as simplify the review process.

miha-plesko replied:
Hm, I totally forgot that for VirtualBox we need Wrapper that answers to UDP broadcasting. As @gberginc said, lets leave OSv Java compiler untouched for now. I've reverted all the deletions.

ilevine commented on pkg/compilers/interface.go:
```diff
@@ -16,6 +16,10 @@ type Compiler interface {
 type CompilerUsage struct {
+	// GeneralDescription section briefly describes what user has
+	// to prepare.
+	GeneralDescription string
```
changes like this should be broken into separate pull requests

miha-plesko replied:
Temptation to implement this tiny upgrade here inline was simply to strong😄
Thanks for pointing out, let me do this properly this time.

ilevine commented on pkg/compilers/osv/osv_dynamic.go:
```diff
@@ -0,0 +1,76 @@
+
+// readImageSizeFromManifest parses manifest.yaml and returns image size string, e.g. "10GB"
+// Returns default image size if anything goes wrong
+func readImageSizeFromManifest(projectDir string) string {
```
it is not clear why you have a `readImageSizeFromManifest` function as well as `readImageSizeFromManifestMB`
it's also not recommended to perform multiple reads on the same file (since this function is called in more than one place).
it would make sense to perform one read on the file, and pass the parsed data as an argument to this function where appropriate

miha-plesko replied:
You are correct. I've resolved this by joining the two functions into one and then calling it only once in a build time. I had to extend `CompileImageParams` (from types.params.go) with additional parameter `SizeMB` to do this though.

ilevine commented on pkg/compilers/osv/osv_java.go:
```diff
@@ -1,75 +1,64 @@
 
 type OSvJavaCompiler struct {
 	ImageFinisher ImageFinisher
 }
```
this should be changed to 
```golang
type OSvJavaCompiler struct {
  	ImageFinisher
}
```

ilevine commented on pkg/compilers/osv/osv_node.go:
```diff
@@ -0,0 +1,104 @@
+	content = "runtime: " + runtime + "\n\n" + content
+
+	// - Write.
+	if err = ioutil.WriteFile(filepath, []byte(content), 0644); err != nil {
```
does it make sense to edit a user-created text file in this function?

miha-plesko replied:
Yes. Thing is that this user-created text file COULD contain the programming language stanza (e.g. "nodejs" or "java"), but @ilackarms insisted that it should be possible to set language using --language argument to unify UniK's UX. E.g.:
```
unik build ... --language nodejs
```

So what we do here is we allow user to use --language parameter and then we insert the appropriate line into her configuration file.

NOTE: We do NOT modify the file provided by user, but a clone of it. The modification is performed on the files that are then attached to Docker container.

ilevine reviewed (changes requested):
some of the changes in this PR belong in separate PRs. 

PRs should be small and contain the minimal amount of changes needed to add or improve a single feature. In most cases, changes to a `provider` and a `compiler` should be proposed in separate pull requests.

It is not likely that this PR will be merged as long as it removes Stub/Wrapper from OSv Java compiler. It is a priority of this project to implement a wrapper in all supported unikernels; removing the wrapper here is a step backwards.

miha-plesko reviewed (commented):
Thank you @ilevine and @gberginc for the review. I've applied your comments and prepared a new PR. The new PR is much shorter and only integrates NodeJS and C/C++ support for OSv leaving Java untouched.

Please see here:
https://github.com/emc-advanced-dev/unik/pull/119

ilevine commented:
closing this PR with respect to the new one that was submitted in its place
